exonum/tests/balance_service.rs

```diff
@@ -0,0 +1,260 @@
+            let snapshot = self.blockchain.snapshot();
+            let schema = BalanceSchema::new(snapshot);
+
+            let balance = schema.balance().get();
+            self.ok_response(&serde_json::to_value(balance).unwrap())
+        }
+
+        /// Common processing for transaction-accepting endpoints.
+        fn post_transaction(&self, req: &mut Request) -> IronResult<Response> {
+            match req.get::<bodyparser::Struct<BalanceTransactions>>() {
+                Ok(Some(transaction)) => {
+                    let transaction: Box<Transaction> = transaction.into();
+                    let tx_hash = transaction.hash();
+                    self.channel.send(transaction).map_err(ApiError::from)?;
+                    let json = TransactionResponse { tx_hash };
+                    self.ok_response(&serde_json::to_value(&json).unwrap())
+                }
+                Ok(None) => Err(ApiError::BadRequest("Empty request body".into()))?,
+                Err(e) => Err(ApiError::BadRequest(e.to_string()))?,
+            }
+        }
+    }
+
+    impl Api for BalanceApi {
+        fn wire(&self, router: &mut Router) {
+            let self_ = self.clone();
+            let post_transaction = move |req: &mut Request| self_.post_transaction(req);
+            let self_ = self.clone();
+            let get_count = move |req: &mut Request| self_.get_count(req);
+
+            router.post("/v1/transaction", post_transaction, "post_transaction");
+            router.get("/v1/count", get_count, "get_count");
+        }
+    }
+}
+
+pub mod service {
+    use exonum::blockchain::{ApiContext, Service, TimeoutAdjusterConfig, Transaction,
+                             TransactionSet};
+    use exonum::{encoding, api::Api, messages::RawTransaction};
+    use exonum::crypto::{gen_keypair, Hash};
+    use exonum::storage::{Database, Fork, MemoryDB, Snapshot};
+    use exonum::node::{ExternalMessage, Node, TransactionSend};
+    use exonum::helpers;
+    use iron::Handler;
+    use router::Router;
+    use serde_json::Value;
+
+    use transactions::{BalanceTransactions, TxAddBalance};
+    use api::BalanceApi;
+    use schema::BalanceSchema;
+
+    use std::sync::Arc;
+    use std::thread;
+    use std::time;
+
+    /// Service ID for the `Service` trait.
+    pub const SERVICE_ID: u16 = 1;
+
+    pub struct BalanceService();
+
+    impl Service for BalanceService {
+        fn service_id(&self) -> u16 {
+            SERVICE_ID
+        }
+
+        fn service_name(&self) -> &str {
+            "balance"
+        }
+
+        fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
+            vec![]
+        }
+
+        fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, encoding::Error> {
+            let tx = BalanceTransactions::tx_from_raw(raw)?;
+            Ok(tx.into())
+        }
+
+        fn public_api_handler(&self, ctx: &ApiContext) -> Option<Box<Handler>> {
```

Comment:
I'm not quite sure why this redefinition and `BalanceApi` are needed. As far as I see, the transactions in the test are sent directly through the MPSC channel, not HTTP API.

Reply:
Seems like you're right. I'll remove it.